### PR TITLE
fix attachments memory consumption

### PR DESCRIPTION
**Pull request by svituz**
### Description:

This PR improves performance when getting attachments of big negotiations. It adds Blaze Persistence dependency since it gives the possibility, compared to JPA projection, to specify a subset of columns also for nested projections.
The authorization on objects (negotiation, attachments, and posts) is checked directly by querying the database instead of fetching unnecessary objects from the DB. 
This should improve memory usage.

### Checklist:

_Make sure you tick all the boxes below if they are true or do not apply:_

- [x] I have performed a self review of my code
- [x] My code follows [Google Java Code style](https://google.github.io/styleguide/javaguide.html)
- [x] I have made my code as simple as possible
- [x] I have added unit tests and the code coverage has not decreased
- [x] I have updated the documentation in all relevant places

### CODE REVIEW

File: src/main/java/eu/bbmri_eric/negotiator/configuration/BlazePersitenceConfig.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.EntityManagerFactory;
+import jakarta.persistence.PersistenceUnit;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.context.annotation.Scope;
+
+@Configuration
+public class BlazePersitenceConfig {
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/configuration/BlazePersitenceConfig.java:17]: typo

File: src/main/java/eu/bbmri_eric/negotiator/database/model/Attachment.java [hunk 2]
```diff
@@ -45,7 +35,7 @@ public class Attachment extends AuditEntity {
   @Column(name = "id")
   private String id;
 
-  @ManyToOne
+  @ManyToOne(fetch = FetchType.LAZY)
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/model/Attachment.java:38]: lazy is default state

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java [hunk 3]
```diff
@@ -27,4 +27,53 @@ Optional<NegotiationResourceState> findNegotiationResourceStateById(
...
+              + "FROM request rq JOIN request_resources_link rrl on rq.id = rrl.request_id "
+              + "                JOIN resource rs on rs.id = rrl.resource_id "
+              + "WHERE rq.negotiation_id = :negotiationId AND "
+              + "      rs.id in ("
+              + "         SELECT  rrl.resource_id "
+              + "         from person p join resource_representative_link rrl ON p.id = rrl.person_id "
+              + "         where p.id = :personId"
+              + "))",
+      nativeQuery = true)
+  boolean isRepresentativeOfAnyResource(String negotiationId, Long personId);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java:43]: this method is not used

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java [hunk 4]
```diff
@@ -27,4 +27,53 @@ Optional<NegotiationResourceState> findNegotiationResourceStateById(
...
+      value =
+          "SELECT distinct(n.id) "
+              + "FROM negotiation n "
+              + "    JOIN request rq ON n.id = rq.negotiation_id "
+              + "    JOIN request_resources_link rrl ON rrl.request_id = rq.id "
+              + "    JOIN resource rs ON rrl.resource_id = rs.id "
+              + "    JOIN organization o ON rs.organization_id = o.id "
+              + "WHERE n.id = :negotiationId and o.external_id = :organizationExternalId",
+      nativeQuery = true)
+  String negotiationWhereOrganization(String negotiationId, String organizationExternalId);
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java:78]: not used either

File: pom.xml [hunk 5]
```diff
@@ -354,8 +355,56 @@
...
             <version>3.1.4</version>
         </dependency>
-
-
+<!--        <dependency>-->
+<!--            <groupId>com.blazebit</groupId>-->
+<!--            <artifactId>blaze-persistence-core-api</artifactId>-->
+<!--            <scope>compile</scope>-->
+<!--            <version>1.6.11</version>-->
+<!--        </dependency>-->
```
Comment (RadovanTomik) [inline pom.xml:363]: please remove commented dep

File: pom.xml [hunk 6]
```diff
@@ -354,8 +355,56 @@
...
+            <groupId>com.blazebit</groupId>
+            <artifactId>blaze-persistence-integration-spring-data-3.1</artifactId>
+            <version>${blaze-persistence.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.blazebit</groupId>
+            <artifactId>blaze-persistence-integration-hibernate-6.2</artifactId>
+            <version>${blaze-persistence.version}</version>
+            <scope>runtime</scope>
+        </dependency>
```
Comment (RadovanTomik) [inline pom.xml:407]: are all these dependencies really needed? Isnt there a parent library containing all of them if yes?

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
Please see comments, only minor things

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java (hunk 4 above)
Comment (svituz) [reply src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java:78]: removed

File: src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java (hunk 3 above)
Comment (svituz) [reply src/main/java/eu/bbmri_eric/negotiator/database/repository/NegotiationRepository.java:43]: removed

File: src/main/java/eu/bbmri_eric/negotiator/database/model/Attachment.java (hunk 2 above)
Comment (svituz) [reply src/main/java/eu/bbmri_eric/negotiator/database/model/Attachment.java:38]: removed

File: src/main/java/eu/bbmri_eric/negotiator/service/DBAttachmentService.java [hunk 7]
```diff
@@ -26,53 +28,93 @@
...
-    this.negotiationService = negotiationService;
-    this.personRepository = personRepository;
-  }
+  @Autowired private AttachmentRepository attachmentRepository;
+  @Autowired private ModelMapper modelMapper;
+  @Autowired private NegotiationRepository negotiationRepository;
+  @Autowired private OrganizationRepository organizationRepository;
+  @Autowired private PersonService personService;
+  @Autowired private NegotiationService negotiationService;
+  @Autowired private PersonRepository personRepository;
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DBAttachmentService.java:37]: please remove the autowired annotations and use a classic constructor

File: src/main/java/eu/bbmri_eric/negotiator/service/DBAttachmentService.java [hunk 8]
```diff
@@ -98,103 +140,71 @@ private AttachmentMetadataDTO saveAttachment(
...
-  private void checkAuthorization(Long userId, Negotiation negotiation) {
-    Person uploader =
-        personRepository.findById(userId).orElseThrow(() -> new EntityNotFoundException(userId));
-    if (!uploader.isAdmin()
-        && !uploader.equals(negotiation.getCreatedBy())
-        && negotiation.getResources().stream().noneMatch(uploader.getResources()::contains)) {
+  private void checkAuthorization(Long userId, String negotiationId, String organizationExternalId)
+      throws ForbiddenRequestException {
+    // if the recipient organization is specified and the sender is not the negotiation creator,
+    // checks whether the user is representative of the organization
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DBAttachmentService.java:146]: please remove comments and rather extract the logic into a separate method with an appropriate name

File: src/main/java/eu/bbmri_eric/negotiator/service/DBAttachmentService.java [hunk 9]
```diff
@@ -98,103 +140,71 @@ private AttachmentMetadataDTO saveAttachment(
...
       // creator of the attachment
-      return attachment.isCreator(
-          NegotiatorUserDetailsService.getCurrentlyAuthenticatedUserInternalId());
+      return isCurrentAuthenticatedUserAttachmentCreator(attachment);
     } else {
       // otherwise the user has to be authorized for the negotiation and
       // the attachment must be either:
       // 1. public (in the negotiation)
       // 2. created by the currently authenticated user
       // 3. addressed to the organization represented by the authenticated user
```
Comment (RadovanTomik) [inline src/main/java/eu/bbmri_eric/negotiator/service/DBAttachmentService.java:191]: same here

**Review by RadovanTomik — CHANGES_REQUESTED** [review]
just some small issues

**Review by RadovanTomik — APPROVED** [review]

**Pull request merged**